libc/dns/getservbyname.c

```diff
@@ -28,33 +28,30 @@
 #include "libc/dns/ent.h"
 #include "libc/dns/servicestxt.h"
 #include "libc/mem/mem.h"
+#include "libc/str/str.h"
 
 struct servent *getservbyname(const char *name, const char *proto) {
   static struct servent *ptr0, se0;
   static char s_name[DNS_NAME_MAX + 1];
-  char *localproto = proto;
+  static char localproto[DNS_NAME_MAX + 1];
   int p;
 
   if (!ptr0) {
     se0.s_name = s_name;
     if (!(se0.s_aliases = calloc(1, sizeof(char *)))) return NULL;
     se0.s_port = 0;
-    se0.s_proto = NULL;
+    se0.s_proto = localproto;
     ptr0 = &se0;
   }
 
-  p = LookupServicesByName(name, &localproto, ptr0->s_name, DNS_NAME_MAX, NULL);
-  if (p == -1) {
-    // localproto got alloc'd during the lookup?
-    if (!proto && localproto != proto) free(localproto);
-    return NULL;
-  }
+  localproto[0] = '\0';
+  if (proto) strncpy(localproto, proto, DNS_NAME_MAX);
```

Comment:
This line isn't memory safe. You'd want to do something like this:

```c
if (proto) {
  if (strlen(proto) > DNS_NAME_MAX) return 0;
  strcpy(localproto, proto);
} else {
  strcpy(localproto, "");
}
```

Update other file too please.